src/OpenTelemetry/Metrics/MeterSdk.cs

```diff
@@ -75,7 +98,27 @@ public void Collect()
                         var aggregator = handle.Value.GetAggregator();
                         aggregator.Checkpoint();
                         this.metricProcessor.Process(this.meterName, metricName, labelSet, aggregator);
+
+                        if (handle.Value.Status == RecordStatus.CandidateForRemoval)
+                        {
+                            boundInstrumentsToRemove.Add(labelSet);
+                        }
+
+                        if (handle.Value.Status == RecordStatus.UpdatePending)
+                        {
+                            handle.Value.Status = RecordStatus.CandidateForRemoval;
+                        }
+                    }
+
+                    foreach (var boundInstrumentToRemove in boundInstrumentsToRemove)
+                    {
+                        // This actual unbinding or removal of the record occurs inside UnBind
+                        // which synchronizes with Bind to ensure no record with pending update
+                        // is lost.
+                        counterInstrument.UnBind(boundInstrumentToRemove);
```

Comment:
Same as for longCounters, could we call `Unbind` here instead of using a temp list?

Reply:
Calling Unbind would cause an entry to be deleted from the dictionary, while were are iterating over the dictionary. Thats why I collected the list, and called UnBind outside the iteration.